vaadin-spreadsheet-flow-parent/vaadin-spreadsheet-flow/src/main/resources/META-INF/resources/frontend/vaadin-spreadsheet/vaadin-spreadsheet.js

```diff
@@ -196,15 +196,31 @@ export class VaadinSpreadsheet extends LitElement {
     super.connectedCallback()
     this.observer && this.observer.observe(this);
     // Restore styles in the case widget is reattached, it happens e.g in client router
-    this.styles && this.styles.forEach(e => document.head.appendChild(e));
+    let i = 0;
+    this.styles && this.styles.forEach(e => {
+      let addedNode = document.head.appendChild(e);
+      if (this.removedRules) {
+        let rulesToBeAdded = this.removedRules[i];
+        for (let j = 0; j<rulesToBeAdded.length ; j++) {
+          let rule = rulesToBeAdded.item(j);
+          addedNode.sheet.insertRule(rule.cssText);
+        }
+        i++;
+      }
+    });
+    this.removedRules = undefined;
   }
 
   disconnectedCallback() {
     super.disconnectedCallback();
     this.observer && this.observer.unobserve(this);
     // Remove styles added to the head by the Widget
     this.styles = document.head.querySelectorAll(`style[id^="spreadsheet-${this.id}"]`);
-    this.styles.forEach(e => document.head.removeChild(e));
+    this.removedRules = [];
+    this.styles.forEach(e => {
+      this.removedRules.push(e.sheet.cssRules);
```

Comment:
Let's avoid adding `this.removedRules` array and store the removed rules with the style element: `e.__removedRules = e.sheet.cssRules;`